### PR TITLE
proposal remove: deserializeSolids and serializeSolids

### PR DESCRIPTION
**Pull request by hrgdavor**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?

I just tried to remove JSON.stringify and JSON.parse from communication with worker.
It seems unnecessary, and all test passed. Also some sample models worked just fine.

Use this version of the PR is to quick change to test the proposal. And if agreed then next step would be to remove deserializeSolids and serializeSolids modules completely.

### CODE REVIEW

**Comment by z3dev** [comment]
@hrgdavor Interesting. I'm sure objects are transferred via JSON automatically, i.e. sending messages with a general payload just does JSON. Can you explore this further?

Also, is there anyway to compare the performance?

**Comment by z3dev** [comment]
FYI, V1 was converting the geometries to compact binary form, which is basically a typed array. This was for performance reasons only.

The down side of using typed arrays is that normal attributes like ‘id’ or ‘name’ were excluded. Users should be able to add attributes to geometries for use in other libraries, etc.

**Comment by hrgdavor** [comment]
stringify and parse are not used behind the scenes, but some optimized  "Structured Clone algorithm" that should be a bit faster, and also handles well circular references unlike JSON.

Someone else did the digging, and it seems that there is no benefit in doing JSON.stringify + parse. Performance difference is hard to measure.
https://surma.dev/things/is-postmessage-slow/
conclusion is that stringify is not worth it

**Comment by hrgdavor** [comment]
> FYI, V1 was converting the geometries to compact binary form, which is basically a typed array. This was for performance reasons only.
> 
> The down side of using typed arrays is that normal attributes like ‘color’ or ‘id’ or ‘name’ were excluded. Users should be able to add attributes to geometries for use in other libraries, etc.

This could have been avoided by only converting numeric array data to binary form, and shallow copy all other properties of each geometry. 
However this approach would, I think, only bring down performance (but one can never assume about performance :) but test instead ).

If on the the other hand geometries would use ArrayBuffers internally (assumming previously mentioned problems are solved),
postMessage has a second argument where you put references to all buffers. In that case those bufers are transfered superfast, ownership changes to the receiver of the postMessage, and code that sends data loses access to them unless later returned back.

I think tat even better approach would be to run both model script execution and rendering in the same worker, and that way there is no need for postMessage and any conversion, they could just share data. I am working on this for the new web viewer.

**Comment by z3dev** [comment]
> I think tat even better approach would be to run both model script execution and rendering in the same worker, and that way there is no need for postMessage and any conversion, they could just share data. I am working on this for the new web viewer.

😎 but be aware that there will be others that want to see just the results for the execution. There’s already an option for style of serialization. 😃

**Comment by hrgdavor** [comment]
> > I think tat even better approach would be to run both model script execution and rendering in the same worker, and that way there is no need for postMessage and any conversion, they could just share data. I am working on this for the new web viewer.
> 
> 😎 but be aware that there will be others that want to see just the results for the execution. There’s already an option for style of serialization. 😃

Yes, execution and render code are separate and would remain separated. So, that use case would not be affected.

Viewer can use postMessage or pass geometries directly to renderer depending if they are in the same thread or not. 

The option for style of serialization is ok, but the version that uses JSON.stringify is likely unnecessary.

serializeSolids would then be ok to keep as it reformats geometries, but deserializeSolids would be obsolete.

**Comment by z3dev** [comment]
@hrgdavor I tested these changes, and did not find any negative effects. I tried every design that I have, plus all previous test cases for previous issues. No problem.

So, please make the final changes. These too functions are a tight pair, so let's remove the JSON and just pass the objects straight through.

P.S. I read the article on post message. Very interesting and in depth. The best solution is just pass the objects, and let the java engines do the serialization. Fun reading!

**Comment by hrgdavor** [comment]
Yeah, the devs working on the javascript engines are doing some great stuff. And the article was fun to read indeed. 
Although I am here because of my 3d printing hobby, I have learned a few nice details about javascript that I will be able to use at work.

File: packages/web/src/ui/flow/design.js [hunk 1]
```diff
@@ -478,21 +477,6 @@ const actions = ({ sources }) => {
...
-        try {
-          const start = new Date().getTime()
-
-          const solids = deserializeSolids(event.data.solids)
-
-          const end = new Date().getTime()
-          console.warn(`elapse for solid generation: ${end - start}`)
-
-          const { lookupCounts, lookup } = event.data
-          return { solids, lookup, lookupCounts }
```
Comment (z3dev) [inline packages/web/src/ui/flow/design.js:491]: The return from this map((event) is different now. Does this work?

File: packages/core/src/code-evaluation/serializeSolids.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ const serializeSolids = (solids) => {
     if (geom2.isA(object)) geom2.toSides(object)
     if (geom3.isA(object)) geom3.toPolygons(object)
     if (path2.isA(object)) path2.toPoints(object)
-    return JSON.stringify(object)
+    return object
```
Comment (z3dev) [inline packages/core/src/code-evaluation/serializeSolids.js:17]: Let's add another NOTE... something like...

NOTE: Serialization of objects is the fastest due to optimizations in the JavaScript engines.

**Review by z3dev — COMMENTED** [review]
Some questions..

**Comment by z3dev** [comment]
@hrgdavor Please see the comments. I like the changes, but let's document WHY this was changed.

File: packages/web/src/ui/flow/design.js (hunk 1 above)
Comment (hrgdavor) [reply packages/web/src/ui/flow/design.js:491]: you are right, it stopped working when this was completely removed. I have made a fix

File: packages/core/src/code-evaluation/serializeSolids.js (hunk 2 above)
Comment (hrgdavor) [reply packages/core/src/code-evaluation/serializeSolids.js:21]: I have adeed a note. Is this ok now  for merging ? :)

**Review by z3dev — APPROVED** [review]
OK. These changes look good.

We don't have automated testing for the WEB UI, so please take extra time to manually test related functionality.

**Pull request merged**